cpp/include/cudf/join.hpp

```diff
@@ -367,5 +367,140 @@ std::unique_ptr<cudf::table> cross_join(
   cudf::table_view const& right,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
 
+/**
+ * @brief Hash join that builds hash table in creation and probes results in subsequent `*_join`
+ * member functions.
+ *
+ * This class enables the hash join scheme that builds hash table once, and probes as many times as
+ * needed (possibly in parallel).
+ */
+class hash_join {
+ public:
+  hash_join() = delete;
+  ~hash_join();
+  hash_join(hash_join const&) = delete;
+  hash_join(hash_join&&)      = delete;
+  hash_join& operator=(hash_join const&) = delete;
+  hash_join& operator=(hash_join&&) = delete;
+
+  /**
+   * @brief Construct a hash join object for subsequent probe calls.
+   *
+   * @note The `hash_join` object must not outlive the table viewed by `build`, else behavior is
+   * undefined.
+   *
+   * @param build The build table, from which the hash table is built.
+   * @param build_on The column indices from `build` to join on.
+   */
+  hash_join(cudf::table_view const& build, std::vector<size_type> const& build_on);
+
+  /**
+   * @brief Side of the probe table in the joined table. Only applicable for inner join.
+   */
+  enum class probe_output_side { LEFT, RIGHT };
```

Comment:
If we do end up keeping this, I think we can improve the docs to make it a bit easier to understand. 
```suggestion
  enum class probe_output_side { 
     LEFT, ///< Columns from the probe table appear as the leftmost in the output table
     RIGHT ///< Columns from the probe table appear as the rightmost in the output table 
  };
```